### PR TITLE
Refactor software renderer

### PR DESCRIPTION
**Pull request by GPUCode**
The software renderer was the first backend that citra had and due to that the code design is quite different than the opengl renderer. It relies on global state and has different namespace naming conventions compared to the hardware renderers. This PR aims to bring the code closer to the standards of the codebase nowadays and tackle most of its technical debt by removing global state and properly encapsulating its functions.

The biggest change is in the rasterizer. Triangle processing has been moved there and split into multiple functions to make it easier to understand. lambdas have been made lower case and variables are made const when possible. The PR also improves software presentation by commonizing the framebuffer logic to the renderer and by ensuring proper synchronization. Previously the Qt/SDL threads were accessing main memory without any sort of guarantees.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6621)

### CODE REVIEW

File: src/video_core/pica_types.h [hunk 1]
```diff
@@ -73,70 +77,70 @@ struct Float {
...
-    bool operator!=(const Float<M, E>& flt) const {
+    constexpr bool operator!=(const Float<M, E>& flt) const {
         return ToFloat32() != flt.ToFloat32();
     }
 
 private:
-    static const unsigned MASK = (1 << (M + E + 1)) - 1;
-    static const unsigned MANTISSA_MASK = (1 << M) - 1;
-    static const unsigned EXPONENT_MASK = (1 << E) - 1;
+    static const u32 MASK = (1 << (M + E + 1)) - 1;
```
Comment (SachinVin) [inline src/video_core/pica_types.h:141]: these can probably be constexpr too

File: src/video_core/renderer_software/renderer_software.cpp [hunk 2]
```diff
@@ -2,18 +2,87 @@
...
 
+#include "common/color.h"
+#include "core/core.h"
+#include "core/hw/gpu.h"
+#include "core/hw/hw.h"
+#include "core/hw/lcd.h"
 #include "video_core/renderer_software/renderer_software.h"
 
-namespace VideoCore {
+namespace Software {
```
Comment (SachinVin) [inline src/video_core/renderer_software/renderer_software.cpp:12]: Honestly, just "Software" is too generic of a term to be used here

Comment (GPUCode) [reply src/video_core/renderer_software/renderer_software.cpp:12]: Recommandations for an alternative name would be appreciated then, not sure what else we can call it

Comment (Steveice10) [reply src/video_core/renderer_software/renderer_software.cpp:12]: Would something like "SoftwareRenderer" work or is that too wordy?

Comment (GPUCode) [reply src/video_core/renderer_software/renderer_software.cpp:12]: Yeah it's a tad too long to be pleasing. Hm perhaps something like SwRenderer is the best of both worlds

File: src/video_core/renderer_software/sw_texturing.cpp [hunk 3]
```diff
@@ -9,41 +9,39 @@
...
-
-    case TexturingRegs::TextureConfig::MirroredRepeat: {
-        unsigned int coord = (static_cast<unsigned>(val) % (2 * size));
+    case TextureConfig::Repeat2:
+    case TextureConfig::Repeat3:
+    case TextureConfig::Repeat:
+        return static_cast<s32>(static_cast<u32>(val) % size);
+    case TextureConfig::MirroredRepeat: {
+        u32 coord = (static_cast<u32>(val) % (2 * size));
         if (coord >= size)
```
Comment (SachinVin) [inline src/video_core/renderer_software/sw_texturing.cpp:41]: braces.. we should add the rule to clang-format

File: src/video_core/renderer_software/sw_texturing.cpp [hunk 4]
```diff
@@ -9,41 +9,39 @@
...
-    case TexturingRegs::TextureConfig::ClampToEdge2:
+    case TextureConfig::ClampToEdge2:
         // For negative coordinate, ClampToEdge2 behaves the same as Repeat
         if (val < 0) {
-            return static_cast<int>(static_cast<unsigned>(val) % size);
+            return static_cast<s32>(static_cast<u32>(val) % size);
         }
-    // [[fallthrough]]
-    case TexturingRegs::TextureConfig::ClampToEdge:
+        [[fallthrough]];
```
Comment (SachinVin) [inline src/video_core/renderer_software/sw_texturing.cpp:25]: nit no semicolon

File: src/video_core/renderer_software/sw_rasterizer.cpp [hunk 5]
```diff
@@ -2,15 +2,943 @@
...
+
+        const auto dstfactor = Common::MakeVec(
+            lookup_factor(0, params.factor_dest_rgb), lookup_factor(1, params.factor_dest_rgb),
+            lookup_factor(2, params.factor_dest_rgb), lookup_factor(3, params.factor_dest_a));
+
+        blend_output = EvaluateBlendEquation(combiner_output, srcfactor, dest, dstfactor,
+                                             params.blend_equation_rgb);
+        blend_output.a() = EvaluateBlendEquation(combiner_output, srcfactor, dest, dstfactor,
+                                                 params.blend_equation_a)
+                               .a();
```
Comment (SachinVin) [inline src/video_core/renderer_software/sw_rasterizer.cpp:658]: clang-format why!?

File: src/video_core/renderer_software/sw_proctex.cpp [hunk 6]
```diff
@@ -112,8 +114,8 @@ static void ClampCoord(float& coord, ProcTexClamp mode) {
     }
 }
 
-static float CombineAndMap(float u, float v, ProcTexCombiner combiner,
```
Comment (SachinVin) [inline src/video_core/renderer_software/sw_proctex.cpp:115]: why is static removed here?

Comment (GPUCode) [reply src/video_core/renderer_software/sw_proctex.cpp:115]: Cause the functions have been placed in an anonymous namespace

File: src/video_core/renderer_software/sw_framebuffer.cpp [hunk 7]
```diff
@@ -99,19 +112,18 @@ const Common::Vec4<u8> GetPixel(int x, int y) {
     return {0, 0, 0, 0};
 }
 
-u32 GetDepth(int x, int y) {
-    const auto& framebuffer = g_state.regs.framebuffer.framebuffer;
+u32 Framebuffer::GetDepth(int x, int y) const {
+    const auto& framebuffer = regs.framebuffer;
     const PAddr addr = framebuffer.GetDepthBufferPhysicalAddress();
-    u8* depth_buffer = VideoCore::g_memory->GetPhysicalPointer(addr);
```
Comment (SachinVin) [inline src/video_core/renderer_software/sw_framebuffer.cpp:105]: I think it makes sense to keep this variable but move it closer to where it is used

File: src/video_core/renderer_software/sw_framebuffer.cpp [hunk 8]
```diff
@@ -155,33 +165,29 @@ u8 GetStencil(int x, int y) {
     }
 }
 
-void SetDepth(int x, int y, u32 value) {
-    const auto& framebuffer = g_state.regs.framebuffer.framebuffer;
+void Framebuffer::SetDepth(int x, int y, u32 value) const {
+    const auto& framebuffer = regs.framebuffer;
     const PAddr addr = framebuffer.GetDepthBufferPhysicalAddress();
-    u8* depth_buffer = VideoCore::g_memory->GetPhysicalPointer(addr);
```
Comment (SachinVin) [inline src/video_core/renderer_software/sw_framebuffer.cpp:161]: ditto

File: src/video_core/renderer_software/sw_framebuffer.cpp (hunk 7 above)
Comment (SachinVin) [inline src/video_core/renderer_software/sw_framebuffer.cpp:105]: I think it makes sense to keep this variable but move it closer to where it is used

File: src/video_core/renderer_software/sw_framebuffer.cpp [hunk 9]
```diff
@@ -190,30 +196,27 @@ void SetDepth(int x, int y, u32 value) {
     }
 }
 
-void SetStencil(int x, int y, u8 value) {
-    const auto& framebuffer = g_state.regs.framebuffer.framebuffer;
+void Framebuffer::SetStencil(int x, int y, u8 value) const {
+    const auto& framebuffer = regs.framebuffer;
     const PAddr addr = framebuffer.GetDepthBufferPhysicalAddress();
-    u8* depth_buffer = VideoCore::g_memory->GetPhysicalPointer(addr);
```
Comment (SachinVin) [inline src/video_core/renderer_software/sw_framebuffer.cpp:196]: ditto

File: src/video_core/renderer_software/sw_framebuffer.cpp [hunk 10]
```diff
@@ -222,36 +225,61 @@ void SetStencil(int x, int y, u8 value) {
...
+    u32 bytes_per_pixel = 4;
+    u32 dst_offset = VideoCore::GetMortonOffset(x, y, bytes_per_pixel) +
+                     coarse_y * framebuffer.width * bytes_per_pixel;
+    u8* dst_pixel = memory.GetPhysicalPointer(addr) + dst_offset;
+
+    auto ref = DecodeD24S8Shadow(dst_pixel);
+    u32 ref_z = ref.x;
+    u32 ref_s = ref.y;
+
+    if (depth < ref_z) {
```
Comment (SachinVin) [inline src/video_core/renderer_software/sw_framebuffer.cpp:245]: can be inverted to return early

File: src/citra/emu_window/emu_window_sdl2_sw.cpp [hunk 11]
```diff
@@ -85,51 +83,22 @@ void EmuWindow_SDL2_SW::Present() {
         };
 
         draw_screen(0);
-        draw_screen(1);
+        draw_screen(2);
```
Comment (SachinVin) [inline src/citra/emu_window/emu_window_sdl2_sw.cpp:86]: why is this 2?

File: src/video_core/renderer_software/sw_texturing.cpp [hunk 12]
```diff
@@ -187,11 +161,11 @@ Common::Vec3<u8> ColorCombine(TevStageConfig::Operation op, const Common::Vec3<u
         // indicate that the per-component computation can't have a higher precision than 1/256,
         // while dot3_rgb((0x80,g0,b0), (0x7F,g1,b1)) and dot3_rgb((0x80,g0,b0), (0x80,g1,b1)) give
         // different results.
-        int result = ((input[0].r() * 2 - 255) * (input[1].r() * 2 - 255) + 128) / 256 +
+        s32 result = ((input[0].r() * 2 - 255) * (input[1].r() * 2 - 255) + 128) / 256 +
                      ((input[0].g() * 2 - 255) * (input[1].g() * 2 - 255) + 128) / 256 +
                      ((input[0].b() * 2 - 255) * (input[1].b() * 2 - 255) + 128) / 256;
         result = std::max(0, std::min(255, result));
```
Comment (SachinVin) [inline src/video_core/renderer_software/sw_texturing.cpp:167]: can be std::clamp

File: src/citra/emu_window/emu_window_sdl2_sw.cpp (hunk 11 above)
Comment (GPUCode) [reply src/citra/emu_window/emu_window_sdl2_sw.cpp:86]: The input now refers to fb_id. 0 is for the left top screen, 1 is for the right top screen and 2 is the bottom screen. This notation is carried over from the gl renderer

File: src/video_core/renderer_software/sw_proctex.cpp (hunk 6 above)
Comment (SachinVin) [reply src/video_core/renderer_software/sw_proctex.cpp:115]: oh okay

File: src/citra/emu_window/emu_window_sdl2_sw.cpp (hunk 11 above)
Comment (SachinVin) [reply src/citra/emu_window/emu_window_sdl2_sw.cpp:86]: can it be an enum then?

File: src/video_core/renderer_software/renderer_software.cpp [hunk 13]
```diff
@@ -2,18 +2,87 @@
...
 
 RendererSoftware::~RendererSoftware() = default;
 
 void RendererSoftware::SwapBuffers() {
+    PrepareRendertarget();
     EndFrame();
 }
 
-} // namespace VideoCore
+void RendererSoftware::PrepareRendertarget() {
```
Comment (FearlessTobi) [inline src/video_core/renderer_software/renderer_software.cpp:25]: nit:
```suggestion
void RendererSoftware::PrepareRenderTarget() {
```

File: src/video_core/renderer_software/renderer_software.cpp [hunk 14]
```diff
@@ -2,18 +2,87 @@
...
+    for (int i : {0, 1, 2}) {
+        int fb_id = i == 2 ? 1 : 0;
```
Comment (FearlessTobi) [inline src/video_core/renderer_software/renderer_software.cpp:27]:
```suggestion
        const int fb_id = i == 2 ? 1 : 0;
```
also please check the other variables if they can be const, I won't comment on it again here

File: src/video_core/renderer_software/sw_clipper.h [hunk 15]
```diff
@@ -1,19 +1,86 @@
...
+        /// TODO: Rounding here is necessary to prevent garbage pixels at
+        ///       triangle borders. Is it that the correct solution, though?
+        return Fix12P4(static_cast<u16>(round(flt.ToFloat32() * 16.0f)));
+    }
+
+    static u16 FracMask() {
+        return 0xF;
+    }
+    static u16 IntMask() {
+        return (u16)~0xF;
```
Comment (FearlessTobi) [inline src/video_core/renderer_software/sw_clipper.h:33]: we do not use c-style casts

File: src/video_core/renderer_software/sw_clipper.h [hunk 16]
```diff
@@ -1,19 +1,86 @@
...
-struct OutputVertex;
+struct TexturingRegs;
 }
 
-namespace Clipper {
+namespace Software {
+
+using Pica::float24;
+
+struct Fix12P4 {
```
Comment (FearlessTobi) [inline src/video_core/renderer_software/sw_clipper.h:19]: Maybe this should have a short comment saying what it does?
I cannot figure it out from the name of the struct

File: src/video_core/renderer_software/sw_framebuffer.cpp [hunk 17]
```diff
@@ -222,36 +225,61 @@ void SetStencil(int x, int y, u8 value) {
...
+        if (stencil == 0) {
+            EncodeD24X8Shadow(depth, dst_pixel);
+        } else {
+            float16 constant = float16::FromRaw(shadow.constant);
+            float16 linear = float16::FromRaw(shadow.linear);
+            float16 x_ = float16::FromFloat32(static_cast<float>(depth) / ref_z);
+            float16 stencil_new = float16::FromFloat32(stencil) / (constant + linear * x_);
+            stencil = static_cast<u8>(std::clamp(stencil_new.ToFloat32(), 0.0f, 255.0f));
+
+            if (stencil < ref_s)
```
Comment (FearlessTobi) [inline src/video_core/renderer_software/sw_framebuffer.cpp:255]: should be braced

File: src/video_core/renderer_software/sw_framebuffer.cpp [hunk 18]
```diff
@@ -222,36 +225,61 @@ void SetStencil(int x, int y, u8 value) {
...
+
+    auto ref = DecodeD24S8Shadow(dst_pixel);
+    u32 ref_z = ref.x;
+    u32 ref_s = ref.y;
+
+    if (depth < ref_z) {
+        if (stencil == 0) {
+            EncodeD24X8Shadow(depth, dst_pixel);
+        } else {
+            float16 constant = float16::FromRaw(shadow.constant);
```
Comment (FearlessTobi) [inline src/video_core/renderer_software/sw_framebuffer.cpp:249]:
```suggestion
            const f16 constant = float16::FromRaw(shadow.constant);
```
?

File: src/video_core/renderer_software/sw_texturing.cpp [hunk 19]
```diff
@@ -9,41 +9,39 @@
 #include "video_core/regs_texturing.h"
 #include "video_core/renderer_software/sw_texturing.h"
 
-namespace Pica::Rasterizer {
+namespace Software {
 
-using TevStageConfig = TexturingRegs::TevStageConfig;
+using TevStageConfig = Pica::TexturingRegs::TevStageConfig;
+
+int GetWrappedTexCoord(Pica::TexturingRegs::TextureConfig::WrapMode mode, int val, unsigned size) {
```
Comment (FearlessTobi) [inline src/video_core/renderer_software/sw_texturing.cpp:16]: I know this is old code, but can we change `unsigned` to `u32`?

**Review by FearlessTobi — APPROVED** [review]
LGTM apart for a few things

File: src/video_core/renderer_software/sw_texturing.cpp (hunk 4 above)
Comment (GPUCode) [reply src/video_core/renderer_software/sw_texturing.cpp:25]: εικόνα

**Comment by GPUCode** [comment]
Renamed float24 to f24 to match with the other types

**Comment by GPUCode** [comment]
Renamed namespace to SwRenderer since there were no more comments

**Review by Steveice10 — APPROVED** [review]
